step02/app.py

```diff
@@ -0,0 +1,84 @@
+from flask import Flask, jsonify, request # web framework
+import boto3 # aws SDK
+
+# openAPI spec
+from openapi_core import create_spec
+from openapi_core.contrib.flask.decorators import FlaskOpenAPIViewDecorator
+import yaml
+
+# DB
+import db
+from models import Images
+
+# misc
+import uuid
+import pathlib
+
+with open('api.yaml') as infile:
+  spec_dict = yaml.load(infile.read())
+  spec = create_spec(spec_dict)
+
+openapi = FlaskOpenAPIViewDecorator.from_spec(spec)
+path = pathlib.PurePath
+
+BUCKET_NAME = 'gg-photo-bucket'
+UPLOAD_FOLDER = 'images/'
+AWS_PROFILE = 'thumbnail-service'
+
+session = boto3.Session(profile_name=AWS_PROFILE)
+s3 = session.client('s3')
+
+app = Flask(__name__)
+
+
+@app.route('/users/me', methods=['GET'])
+@openapi
+def get_profile():
+  return jsonify({
+    'id': 0
+  })
+
+@app.route('/images/upload_url', methods=['GET'])
+@openapi
+def get_upload_url():
+    # query param
+    filename = request.openapi.parameters.query['filename']
+    content_type = request.openapi.parameters.query['content-type']
+    # values used for S3 presigned-url
+    stem = path(filename).stem
+    file_uuid = uuid.uuid4()
+    key = UPLOAD_FOLDER + f'{file_uuid}'
+
+    # upload image in database
+    image_record = Images(bucket=BUCKET_NAME, filename=filename, key=key)
+    db.session.add(image_record)
+    db.session.commit()
```

Comment:
Wrap db operation in try block. If saving to db fails you should probably not generate a presigned url

Reply:
I guess this was bad advise. Usually you want to enumerate all the failures and communicate the error to the API caller so they know how it failed.

Reply:
What I recommend is to preemptively identify 100% of the known error scenarios during the planning phase. You want to implement these errors as error classes and maintain them in a dedicated location. 

[example 1: werkzeug](https://github.com/pallets/werkzeug/blob/1.0.1/src/werkzeug/exceptions.py)
[example 2: psycopg2](https://github.com/psycopg/psycopg2/blob/2_8_5/lib/errorcodes.py)
[example 3: python-arango](https://github.com/joowani/python-arango/blob/5.4.0/arango/exceptions.py)

This is because the errors are also a part of your API's developer interface. The errors thrown from your API should be well defined and documented. The advice I gave is bad because what I would like to see is involved and and outside the scope of this challenge.

What you want to do is enumerate a family of errors and write an [error handler](https://flask.palletsprojects.com/en/1.1.x/errorhandling/) so that the way your API responds to the HTTP request in the case of exceptions is standardized.

For example, if your WebAPI error payload schema is 
```json
{
  "code": 1234,
  "message": "Database connection failure, please resubmit request"
}
```
and your base and example exceptions are:
```python
class MyBaseException(Exception):
    def __repr__(self):
        return f'[{self.code}] self.message'

class MyExampleException(MyBaseException):
    message = "Database connection failure, please resubmit request"
    code = 1234
```
then all errors can be handled like the following

```python
@app.errorhandler(MyBaseException)
def handle_bad_request(e):
    # this will log "[code] message"
    app.logger.error(e)

    # this will make sure payload is always the same
    response = e.get_response()
    response.data = json.dumps({
        "code": e.code,
        "message": e.message,
    })
    response.content_type = "application/json"
    return response
```

Reply:
here's what "enumerating errors" may look like

![image](https://user-images.githubusercontent.com/1236436/85188586-e28c8880-b275-11ea-981c-5ce919f44e57.png)

Reply:
I see. Thank makes sense, I definitely didn't consider that at all. Thanks, for explaining.